sled-storage/src/dataset.rs

```diff
@@ -364,6 +430,352 @@ pub(crate) async fn ensure_zpool_has_datasets(
     Ok(())
 }
 
+#[derive(Debug, thiserror::Error)]
+pub enum DatasetEncryptionMigrationError {
+    #[error(transparent)]
+    IoError(#[from] std::io::Error),
+
+    #[error("Failed to run command")]
+    FailedCommand { command: String, stderr: Option<String> },
+
+    #[error("Cannot create new encrypted dataset")]
+    DatasetCreation(#[from] illumos_utils::zfs::EnsureFilesystemError),
+
+    #[error("Missing stdout stream during 'zfs send' command")]
+    MissingStdoutForZfsSend,
+}
+
+fn status_ok_or_get_stderr(
+    command: &tokio::process::Command,
+    output: &std::process::Output,
+) -> Result<(), DatasetEncryptionMigrationError> {
+    if !output.status.success() {
+        let stdcmd = command.as_std();
+        return Err(DatasetEncryptionMigrationError::FailedCommand {
+            command: format!(
+                "{:?} {:?}",
+                stdcmd.get_program(),
+                stdcmd
+                    .get_args()
+                    .collect::<Vec<_>>()
+                    .join(std::ffi::OsStr::new(" "))
+            ),
+            stderr: Some(String::from_utf8_lossy(&output.stderr).to_string()),
+        });
+    }
+    Ok(())
+}
+
+/// Migrates unencrypted datasets to their encrypted formats.
+pub(crate) async fn ensure_zpool_datasets_are_encrypted(
+    log: &Logger,
+    zpool_name: &ZpoolName,
+) -> Result<(), DatasetEncryptionMigrationError> {
+    info!(log, "Looking for unencrypted datasets in {zpool_name}");
+    let unencrypted_datasets =
+        find_all_unencrypted_datasets_directly_within_pool(&log, &zpool_name)
+            .await?;
+
+    // TODO: Could do this in parallel?
+    for dataset in unencrypted_datasets {
+        let log = &log.new(slog::o!("dataset" => dataset.clone()));
+        info!(log, "Found unencrypted dataset");
+
+        ensure_zpool_dataset_is_encrypted(&log, &zpool_name, &dataset).await?;
+    }
```

Comment:
I'm not sure if we *actually* want to parallelize this, but it would be easy to do.

Seems like we should make disk addition parallel at least, so we can fire up all U.2s at once -- but we might get less benefit from putting extra workload on a U.2 already performing a migration. That's all speculative though.

Reply:
I'm going to start adding some concurrency support later today, or Monday at the latest. Got a touch delayed.

Reply:
All good - I think it would be nice to have, at any time, but I don't think it needs to be a blocker here.